staging/src/k8s.io/kubectl/pkg/cmd/annotate/annotate_test.go

```diff
@@ -731,3 +762,201 @@ func TestAnnotateMultipleObjects(t *testing.T) {
+			resourceVersion: "2",
+			expectObj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations:     map[string]string{"a": "b", "c": "d"},
+					ResourceVersion: "2",
+				},
+			},
+			expectMsg: MsgAnnotationModified,
+		},
+		{
+			obj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{"a": "b"},
+				},
+			},
+			annotations: map[string]string{},
+			remove:      []string{"a"}, // exist
+			expectObj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{},
+				},
+			},
+			expectMsg: MsgAnnotationModified,
+		},
+		{
+			obj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{"a": "b"},
+				},
+			},
+			annotations: map[string]string{},
+			remove:      []string{"c"}, // not exist
+			expectObj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{"a": "b"},
+				},
+			},
+			expectMsg: MsgAnnotationUnchanged,
+		},
+		{
+			obj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{"a": "b", "c": "d"},
+				},
+			},
+			annotations: map[string]string{"e": "f"},
+			remove:      []string{"a"},
+			expectObj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"c": "d",
+						"e": "f",
+					},
+				},
+			},
+			expectMsg: MsgAnnotationModified,
+		},
+		{
+			obj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{"status": "unhealthy"},
+				},
+			},
+			annotations: map[string]string{"status": "healthy"},
+			overwrite:   true,
+			expectObj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"status": "healthy",
+					},
+				},
+			},
+			expectMsg: MsgAnnotationModified,
+		},
+		{
+			obj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{"status": "unhealthy"},
+				},
+			},
+			annotations: map[string]string{"status": "healthy"},
+			overwrite:   false,
+			expectObj: &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						"status": "unhealthy",
+					},
+				},
+			},
+			expectMsg: MsgAnnotationUnchanged,
+			expectErr: true,
+		},
+	}
+
+	for i, test := range tests {
+		options := &AnnotateOptions{
+			overwrite:         test.overwrite,
+			newAnnotations:    test.annotations,
+			removeAnnotations: test.remove,
+			resourceVersion:   test.resourceVersion,
+		}
+
+		oldData, err := json.Marshal(test.obj)
+		if err != nil {
+			t.Errorf("unexpected error: %v %v", err, i)
+		}
+
+		err = options.updateAnnotations(test.obj)
+		if test.expectErr {
+			if err == nil {
+				t.Errorf("unexpected non-error: %v", i)
+			}
+			continue
+		}
+		if !test.expectErr && err != nil {
+			t.Errorf("unexpected error: %v %v", err, i)
+		}
+
+		newData, err := json.Marshal(test.obj)
+		if err != nil {
+			t.Errorf("unexpected error: %v %v", err, i)
+		}
+
+		dataChangeMsg := updateDataChangeMsg(oldData, newData)
```

Comment:
While I appreciate all the test cases here, I would prefer that we make these assertions in a more integrated manner rather than against this very specific function, which is an implementation detail IMO. In other words, I think it would be better for the long term to augment the tests that cover `options.RunAnnotate` to assert on the message that gets printed, so that the tests will continue to be valid if we no longer set the message in this exact same way.

Reply:
Updated.